src/zdns/resolver.go

```diff
@@ -86,6 +88,10 @@ type ResolverConfig struct {
 
 // PopulateAndValidate checks if the ResolverConfig is valid and populates any missing fields with default values.
 func (rc *ResolverConfig) PopulateAndValidate() error {
+	// when creating multiple resolvers, each will call this function against the same config.
+	rc.Lock()
```

Comment:
I don't understand why this needs to be locked. Shouldn't anything for the shared thing be done once ahead of time?

Reply:
So it is. The CLI calls `config.PopulateAndValidate` once and no lock is required from the CLI's point-of-view

However, since someone can use the library without the CLI, inside `InitResolver` we also check that the config is populated with defaults and is valid [here](https://github.com/zmap/zdns/blob/main/src/zdns/resolver.go#L346). When used with the CLI, this is a redundant check. The initialization of workers happens in a multi-threaded fashion [here](https://github.com/zmap/zdns/blob/main/src/cli/worker_manager.go#L325), so the lock is needed for that case.

Reply:
Even if multiple threads are looking at the config at once, I'm not sure I understand why they should have any need to modify the config.

Reply:
That's fair, and as an ideal it should be idempotent when the config is populated/valid to run again.

The issue I saw was around `LocalAddrs`. A single worker would complain that after validation, the `localAddrs` was empty and it would panic.

I think the issue is [here](https://github.com/zmap/zdns/blob/main/src/zdns/resolver.go#L199-L208) where we checked for IPv6 local addresses (since we don't support IPv6 in `main`) and replaced the IPv4 local addresses with all local addresses that pass validation. When called on a valid config, this is just going to replace the `localAddrs` array with an equivalent one, but it's still a write to a shared data structure.

My thinking was I could either try to ensure that a valid config would never be modified at all (even to replace a config field with an equal value as I think is happening here) to ensure idempotency in multi-threaded scenarios, or I just add a lock.

Figured the lock was easier.

Reply:
That's very fair. It's been rather non-trivial to do both
- populate missing values
- verify config is valid
 in the same function correctly and without causing issues in some edge case.
 
Splitting the concerns into CLI/out-side caller **populates** the ResolverConfig and there's a `ValidateConfig` function for `InitResolver` to use to be sure the config is **valid** seems cleaner and we don't have this weirdness.

I'll make the change.